elkm1_lib/connection.py

```diff
@@ -1,287 +1,188 @@
+        self._check_write_queue = asyncio.Event()
+        self._response_received = asyncio.Event()
+        self._heartbeat_event = asyncio.Event()
+        self._tasks: set[asyncio.Task[Any]] = set()
 
     async def connect(self) -> None:
-        """Asyncio connection to Elk."""
+        """Create connection to Elk."""
 
         LOG.info("Connecting to ElkM1 at %s", self._url)
+        retry_time = 1
         scheme, dest, param, ssl_context = parse_url(self._url)
-        heartbeat_time = 120 if scheme != "serial" else -1
-
-        connection = partial(
-            _ElkProtocol,
-            self._loop,
-            heartbeat_time,
-            self._connected_callback,
-            self._disconnected_callback,
-            self._got_data_callback,
-            self._timeout_callback,
-        )
-        try:
+        while not self._writer:
             if scheme == "serial":
-                await serial_asyncio.create_serial_connection(
-                    self._loop, connection, dest, baudrate=param
-                )
+                coro = serial_asyncio.open_serial_connection(url=dest, baudrate=param)
             else:
-                await asyncio.wait_for(
-                    self._loop.create_connection(
-                        connection, host=dest, port=param, ssl=ssl_context
-                    ),
-                    timeout=30,
+                coro = asyncio.open_connection(host=dest, port=param, ssl=ssl_context)
+            try:
+                reader, self._writer = await asyncio.wait_for(coro, timeout=30)
+            except (ValueError, OSError, asyncio.TimeoutError) as err:
+                LOG.warning(
+                    "Error connecting to ElkM1 (%s). Retrying in %d seconds",
+                    err,
+                    retry_time,
                 )
-        except (ValueError, OSError, asyncio.TimeoutError) as err:
-            if self._connection_retry_time <= 0:
-                return
-
-            LOG.warning(
-                "Could not connect to ElkM1 (%s). Retrying in %d seconds",
-                err,
-                self._connection_retry_time,
-            )
-            self._start_connection_retry_timer()
-
-    def is_connected(self) -> bool:
-        """Is the connection active?"""
-        return self._elk_protocol is not None
+                await asyncio.sleep(retry_time)
+                retry_time = min(60, retry_time * 2)
+                continue
+
+            if scheme != "serial":
+                self._tasks.add(asyncio.create_task(self._heartbeat_timer()))
+            self._tasks.add(asyncio.create_task(self._read_stream(reader)))
+            self._tasks.add(asyncio.create_task(self._write_stream()))
+            self._notifier.notify("connected", {})
+
+    async def _read_stream(self, reader: asyncio.StreamReader) -> None:
+        read_buffer = ""
+        while True:
+            data = await reader.read(500)
```

Comment:
Can you use `readuntil` here?

https://docs.python.org/3/library/asyncio-stream.html#asyncio.StreamReader.readuntil

Reply:
This will likely raise https://docs.python.org/3/library/asyncio-exceptions.html#asyncio.IncompleteReadError on disconnect as well

Reply:
See comment on exceptions below.